lib/_http_outgoing.js

```diff
@@ -341,17 +341,22 @@ OutgoingMessage.prototype._send = function _send(data, encoding, callback) {
 OutgoingMessage.prototype._writeRaw = _writeRaw;
 function _writeRaw(data, encoding, callback) {
   const conn = this.socket;
-  if (conn && conn.destroyed) {
-    // The socket was destroyed. If we're still trying to write to it,
-    // then we haven't gotten the 'close' event yet.
-    return false;
-  }
 
   if (typeof encoding === 'function') {
     callback = encoding;
     encoding = null;
   }
 
+  if (conn?.destroyed) {
+    conn._hadError = true;
```

Comment:
I don't think this is necessary

Reply:
I still think this should not be necessary here and if there are problems without it then we have a bug somewhere else.

Reply:
> Calling res.end will trigger onFinshed

How? You have a return inside the `conn?.destroyed` branch.

Reply:
try this:

```js
diff --git a/lib/_http_outgoing.js b/lib/_http_outgoing.js
index 50ef063241..ec2bf87ace 100644
--- a/lib/_http_outgoing.js
+++ b/lib/_http_outgoing.js
@@ -340,18 +340,21 @@ OutgoingMessage.prototype._send = function _send(data, encoding, callback) {
 
 OutgoingMessage.prototype._writeRaw = _writeRaw;
 function _writeRaw(data, encoding, callback) {
+  if (typeof encoding === 'function') {
+    callback = encoding;
+    encoding = null;
+  }
+
   const conn = this.socket;
-  if (conn && conn.destroyed) {
+  if (conn?.destroyed) {
+    if (callback) {
+      process.nextTick(callback, new ERR_STREAM_DESTROYED('write'));
+    }
     // The socket was destroyed. If we're still trying to write to it,
     // then we haven't gotten the 'close' event yet.
     return false;
   }
 
-  if (typeof encoding === 'function') {
-    callback = encoding;
-    encoding = null;
-  }
-
   if (conn && conn._httpMessage === this && conn.writable) {
     // There might be pending data in the this.output buffer.
     if (this.outputData.length) {
@@ -804,9 +807,9 @@ OutgoingMessage.prototype.addTrailers = function addTrailers(headers) {
   }
 };
 
-function onFinish(outmsg) {
-  if (outmsg && outmsg.socket && outmsg.socket._hadError) return;
-  outmsg.emit('finish');
+function onFinish(err) {
+  if (err || this.socket?._hadError) return;
+  this.emit('finish');
 }
 
 OutgoingMessage.prototype.end = function end(chunk, encoding, callback) {
@@ -848,7 +851,7 @@ OutgoingMessage.prototype.end = function end(chunk, encoding, callback) {
   if (typeof callback === 'function')
     this.once('finish', callback);
 
-  const finish = FunctionPrototypeBind(onFinish, undefined, this);
+  const finish = FunctionPrototypeBind(onFinish, this);
 
   if (this._hasBody && this.chunkedEncoding) {
     this._send('0\r\n' + this._trailer + '\r\n', 'latin1', finish);
```

Reply:
> How? You have a return inside the `conn?.destroyed` branch.

1. When `res.end` is called, `onFinish` will be passed in as a `callback`.

https://github.com/nodejs/node/blob/db79783bad754fb659328e53ecd7ab50c8055bf1/lib/_http_outgoing.js#L851-L854

2. If `conn.destroyed` is True, then `onFinish` will be called.
3. `onFinish` will trigger a new `close` event.

Reply:
Please see my previous comment

Reply:
> Please see my previous comment

Updated the code based on previous comments, thanks for the guidance!